language_service/src/identifier_locator.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::mem::replace;
+use std::rc::Rc;
+
+use crate::qsc_utils::{find_ident, find_item, span_contains, span_touches, Compilation};
+use qsc::ast::visit::{walk_expr, walk_namespace, walk_pat, walk_ty_def, Visitor};
+use qsc::{ast, hir, resolve};
+
+#[allow(unused_variables)]
+pub(crate) trait LocatorAPI<'package> {
+    fn at_callable_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        decl: &'package ast::CallableDecl,
+    ) {
+    }
+
+    fn at_callable_ref(
+        &mut self,
+        path: &'package ast::Path,
+        item_id: &'package hir::ItemId,
+        item: &'package hir::Item,
+        package: &'package hir::Package,
+        decl: &'package hir::CallableDecl,
+    ) {
+    }
+
+    fn at_new_type_def(&mut self, type_name: &'package ast::Ident, def: &'package ast::TyDef) {}
+
+    fn at_new_type_ref(
+        &mut self,
+        path: &'package ast::Path,
+        item_id: &'package hir::ItemId,
+        item: &'package hir::Item,
```

Comment:
My comments about `ItemId` and `Item` above for `at_callable_ref` also apply here.